feat(lint): support pnpm catalogs

apple-yagi opened the pull request:
## Summary

resolved #8393

- make `noReactForwardRef` resolve React versions through pnpm catalogs(`catalog:` / `catalog:<name>`) when `javascript.resolver.experimentalPnpmCatalogs` is enabled
- refresh catalog-based resolution when `pnpm-workspace.yaml` changes

## Test Plan

- `cargo test`
- `cargo build --release`
- Verified behavior with a locally built Biome CLI binary:
  - With `javascript.resolver.experimentalPnpmCatalogs: true`, `noReactForwardRef` triggers for React 19 resolved via `catalog:` / `catalog:<name>`
  - With the flag disabled, catalog-based resolution is not applied
  - Editing `pnpm-workspace.yaml` updates diagnostics without restarting
- Verified the same behavior in VS Code by configuring the extension/LSP to use the locally built Biome binary

## Docs

- Updated in-repo docs/comments for opt-in pnpm catalog support

## AI Assistance

- This PR was written with assistance from Codex (ChatGPT).

apple-yagi committed:
fix: handle pnpm catalog for noReactForwardRef

apple-yagi committed:
feat: add support for named catalogs in pnpm workspace

ematipico reviewed (changes requested):
This feels like more a feature than a fix, don't you agree? I am going to block it so you can prepare the PR accordingly:
- changeset for the feature
- **more tests**: we haven't added any tests for the "parsing" of `pnpm-workspace.yml`.
- implement the feature according to our architecture: 
  - use our files system 
  - resolve the pnpm workspace file correctly: it doesn't take into account the root of the project

apple-yagi committed:
fix: update find_pnpm_workspace_catalog to use the current directory as the base

apple-yagi committed:
fix: refactor find_pnpm_workspace_catalog to take a FileSystem trait

apple-yagi committed:
docs: clarify catalog resolution helpers

apple-yagi committed:
refactor: simplify catalog prefix match

apple-yagi committed:
refactor: rename pnpm catalog entry parser

apple-yagi committed:
test: add tests for finding catalog from workspace file

apple-yagi committed:
test: add tests for pnpm workspace catalog parsing and dependency resolution

apple-yagi committed:
feat: add support for pnpm catalogs in dependency resolution for linting

apple-yagi committed:
docs: clarify PackageJson catalog field

apple-yagi commented:
@ematipico I've made corrections.

> more tests: we haven't added any tests for the "parsing" of pnpm-workspace.yml.

My understanding of this comment might be insufficient. Please let me know if there's anything I'm missing.

apple-yagi committed:
fix: normalize pnpm catalog parsing and fs-based tests

dyc3 commented on crates/biome_js_analyze/tests/specs/nursery/noReactForwardRef/react19-catalog/catalog.js:
```diff
@@ -0,0 +1,7 @@
+/* should generate diagnostics with pnpm catalog */
```
These comments are magic comments that tell our infra to make sure these do or don't emit diagnostics, but they need to be exact matches in order for it to work.

```suggestion
/* should generate diagnostics */
```

dyc3 commented on crates/biome_test_utils/src/lib.rs:
```diff
@@ -249,12 +249,32 @@ fn get_js_like_paths_in_dir(dir: &Utf8Path) -> Vec<BiomePath> {
+        let workspace_file = dir.join("pnpm-workspace.yaml");
+        if !fs.path_is_file(&workspace_file) {
+            continue;
+        }
```
For tests, we need them to be as isolated as possible, and going up the file tree all the way searching for files breaks that.

We need to search for the `pnpm-workspace.yaml` that is adjacent to the file being tested. For example, for grabbing `package.json` for a test case, we have:

```
let package_json_file = input_file.with_extension("package.json");
```

Also, I think it needs to support the `yml` file extension too.

siketyan replied:
> Also, I think it needs to support the yml file extension too.

I don't think so, pnpm only supports pnpm-workspace.yaml (exact match): https://pnpm.io/pnpm-workspace_yaml

apple-yagi replied:
I was checking the internal implementation of pnpm, and it seemed to be fixed in pnpm-workspace.yaml.

https://github.com/pnpm/pnpm/blob/98a0410aa1e42784849c5def98341011144a3da4/packages/constants/src/index.ts#L12

dyc3 reviewed (dismissed):
I couldn't find anything that actually _loads_ the `pnpm-workspace.yaml` file at runtime.

There needs to be at least one test in `biome_cli` that shows the file gets read.

apple-yagi committed:
Update crates/biome_js_analyze/tests/specs/nursery/noReactForwardRef/react19-catalog/catalog.js

Co-authored-by: Carson McManus <dyc3@users.noreply.github.com>

apple-yagi committed:
Update crates/biome_js_analyze/tests/specs/nursery/noReactForwardRef/react19-named-catalog/catalog.js

Co-authored-by: Carson McManus <dyc3@users.noreply.github.com>

apple-yagi committed:
refactor: use biome_yaml_parser AST to parse pnpm workspace catalogs

apple-yagi committed:
test: look for pnpm-workspace.yaml next to test input

apple-yagi committed:
feat: load pnpm workspace catalogs and integrate with layout

apple-yagi committed:
feat: refresh package catalogs when pnpm-workspace.yaml changes

apple-yagi committed:
fix: update diagnostic comments in React 19 catalog snapshots

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

apple-yagi committed:
fix lint

apple-yagi committed:
fix: update load_pnpm_workspace_catalog to accept a starting directory

apple-yagi commented on crates/biome_service/src/workspace/server.rs:
```diff
@@ -818,6 +818,28 @@ impl WorkspaceServer {
+        }
+
+        None
+    }
```
fix https://github.com/biomejs/biome/pull/8396/commits/98b6e39e1e544e257ed2d36b0084c5e2bd47d31e

apple-yagi committed:
Update .changeset/nasty-cities-check.md

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

apple-yagi committed:
fix lint

apple-yagi committed:
Update crates/biome_test_utils/src/lib.rs

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

enrique-ramirez commented:
I'm sorry to intrude, but I got bitten by this today and landed here. I hope I'm not messing up, but pinging @dyc3 and @ematipico just in case these fell off your radar and hoping it might be ready to be merged.

ematipico commented:
@enrique-ramirez OP still didn't address all my comments properly. Plus this is a feature that must point to `next`, with a minor changeset. Unfortunately, I can't babysit all PRs. If you need this feature, you can pick it up and complete it, I believe it's almost there

apple-yagi committed:
chore(changeset): switch pnpm catalog changeset from patch to minor

ematipico reviewed (commented):
So, I believe code wise the feature is fine, however I'm a bit concerned about the stability.

The feature relies on the reading of the yaml lock file, which has its own format and conventions, which can change from version to another of pnpm. It's not the first time that happens. If that happens, this feature breaks and Biome doesn't work anymore. 

Have you thought about it? How do you plan to address that?

apple-yagi committed:
test: update pnpm catalog diagnostics for suspicious/noReactForwardRef

apple-yagi commented:
> If possible, I'd prefer to take same road we took with editor config. Opt-in via configuration.

Thanks for the feedback. I’d like to check whether we really need a flag for this.

My reason is that when `pnpm-workspace.yaml` exists, users generally expect it to be loaded automatically.
Requiring users to explicitly enable a flag for that feels redundant. Even if we document it, people may miss it and still open issues.

Loading `pnpm-workspace.yaml` is fail-safe, so even if a breaking change causes a read failure, the result is only that we cannot read the current React version.

So I think we may not need a flag. I will add tests that guarantee fail-safe behavior and add documentation for `pnpm-workspace.yaml`.

That said, I can also implement it as opt-in if you prefer.

ematipico commented:
Sorry I thought we were reading the lock file! 

I have mixed feelings. From one hand I understand your argument @apple-yagi where doing things automatically is best, however we tried the same approach with editorconfig (true by default), and it was a mess DX wise because it turned out that people were using things wrong.

Still, I would like to make opt-in **for now** for two reasons:
- we're using the yaml parser which isn't stable yet, so things can change internally and break the parser 
- we don't know how it will function in the real world. Our tests don't necessarily reflect full fledged projects 

So here my proposal, having a brand new config named `resolver.experimentalCatalogs`

What do you @siketyan ?

apple-yagi commented:
@ematipico Thanks for the background and explanation.
I understand the reasoning now, and it makes sense to me. I’ll switch the implementation to opt-in.

apple-yagi committed:
feat(configuration): add javascript resolver pnpm catalogs opt-in

apple-yagi committed:
test(package): add fail-safe tests for pnpm workspace catalog parsing

apple-yagi committed:
docs(configuration): clarify experimentalPnpmCatalogs behavior

apple-yagi committed:
fix(fs): classify pnpm-workspace.yaml as a manifest file

apple-yagi committed:
fix(lsp): reload workspace on pnpm-workspace.yaml changes

apple-yagi committed:
fix(service): index pnpm-workspace.yaml and refresh catalogs on change

apple-yagi committed:
docs(package): document Catalogs field mapping with YAML examples

apple-yagi committed:
docs(changeset): note pnpm catalog support is opt-in

apple-yagi commented:
I changed the implementation to support pnpm catalogs as an opt-in feature. I also updated it to detect changes to pnpm-workspace.yaml in real time.
I believe all implementation work is now complete, so I’d appreciate a review when you have time.

apple-yagi committed:
Merge branch 'next' into fix/no-react-forward-ref-pnpm-catalog

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes

ematipico committed:
Merge remote-tracking branch 'origin/next' into fix/no-react-forward-ref-pnpm-catalog

ematipico committed:
fix merge conflicts

ematipico committed:
fix merge conflicts pt2

autofix-ci[bot] committed:
[autofix.ci] apply automated fixes
